### PR TITLE
[feat]:[CDS-41438]: Handle responses from the delegate in the manager for AzureCreateTask and BP for NG

### PR DESCRIPTION
**Pull request by nassergonzalez**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 6 Issues 
* Code issue 4 Code issues 
* Anti-pattern 2 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36980)

**Review by tmacari-harness — DISMISSED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: :shipit: complete! ✓

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCommonHelper.java` line 222 at r1](https://reviewable.io/reviews/harness/harness-core/36980#-NAtG7qwCEONhVmnJVQJ:-NAtG7qwCEONhVmnJVQK:bqcfhwx) ([raw file](https://github.com/harness/harness-core/blob/97ce4288c207183b79aa658533d5cd529063207e/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCommonHelper.java#L222)):*
> ```Java
>   }
> 
>   protected String generateIdentifier(String provisionerIdentifier, Ambiance ambiance) {
> ```

why protected

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java [hunk 1]
```diff
@@ -202,8 +225,71 @@ public StepResponse finalizeExecutionWithSecurityContext(Ambiance ambiance, Step
...
+        AzureARMTemplateDataOutput.builder()
+            .resourceGroup(data.getResourceGroup())
+            .resourceGroupTemplateJson(data.getResourceGroupTemplateJson())
+            .subscriptionId(data.getSubscriptionId())
+            .scopeType(scope)
+            .build();
+    String identifier = azureCommonHelper.generateIdentifier(provisionerIdentifier, ambiance);
+    String sweepingOutputKey = format(AZURE_TEMPLATE_DATA_FORMAT, identifier);
+    executionSweepingOutputService.consume(
+        ambiance, sweepingOutputKey, azureARMTemplateDataOutput, StepOutcomeGroup.STAGE.name());
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java:275]: hi Nasser, just wanted to confirm if we need to save based on  StepOutcomeGroup.STAGE and not STEP.

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: :shipit: complete! ✓

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java` line 256 at r1](https://reviewable.io/reviews/harness/harness-core/36980#-NAxX0SbA-aTBv875F_9:-NAxX0SbA-aTBv875F_A:b1o9aj3) ([raw file](https://github.com/harness/harness-core/blob/97ce4288c207183b79aa658533d5cd529063207e/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java#L256)):*
> ```Java
>           "Error while processing Azure Create ARM Resource Task response %s", ex.getCause().getMessage());
>       log.error(errorMsg, ex);
>       return azureCommonHelper.getFailureResponse(ex.getCommandUnitsProgress().getUnitProgresses(), errorMsg);
> ```

Same here, exception should not be handled
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java` line 289 at r1](https://reviewable.io/reviews/harness/harness-core/36980#-NAxX6y09f6JH1FxL6tf:-NAxX6y09f6JH1FxL6tg:b-6zigm9) ([raw file](https://github.com/harness/harness-core/blob/97ce4288c207183b79aa658533d5cd529063207e/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java#L289)):*
> ```Java
>       json.forEach((key, object) -> outputMap.put(key, ((Map<String, Object>) object).get("value")));
>     } catch (IOException exception) {
>       log.error("Exception while parsing ARM outputs", exception);
> ```

nit: if this is not a failure then warn level should be used
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateBPStep.java` line 177 at r1](https://reviewable.io/reviews/harness/harness-core/36980#-NAxWcVeBoYK2iMVNxNc:-NAxWcVeBoYK2iMVNxNd:b-y9gq7l) ([raw file](https://github.com/harness/harness-core/blob/97ce4288c207183b79aa658533d5cd529063207e/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateBPStep.java#L177)):*
> ```Java
>       log.error(errorMsg, ex);
> 
>       return azureCommonHelper.getFailureResponse(ex.getCommandUnitsProgress().getUnitProgresses(), errorMsg);
> ```

exception shouldn't be handled (if no additional handling required, in this case exception should be thrown after) and instead let be handled by framework. In this case I can see that getFailureResponse is handling this and returning StepResponse

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 2 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java` line 256 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBCZPmq2_9hW50lgBs4:-NBCZPmq2_9hW50lgBs5:b-uezz8o) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java#L256)):*
> ```Java
>           "Error while processing Azure Create ARM Resource Task response %s", ex.getCause().getMessage());
>       log.error(errorMsg, ex);
>       throw new RuntimeException(ex.getCause());
> ```

Shouldn't we re-throw it instead of creating a new one?
@abosii-harness 
___
*[`125-cd-nextgen/src/test/resources/recaster/alias/alias.txt` line 466 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBCZoF94jvY4a153P9C:-NBCZoF94jvY4a153P9D:byquo9c) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/test/resources/recaster/alias/alias.txt#L466)):*
> ```help
> io.harness.cdng.provision.azure.AzureCreateARMResourceOutcome
> io.harness.cdng.provision.azure.beans.AzureARMTemplateDataOutput
> io.harness.cdng.infra.beans.host.HostFilter
> ```

Do we need that?

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 2 change requests. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java` line 256 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBCZPmq2_9hW50lgBs4:-NBD8DFgCwGOw-sn2XzB:b28xg56) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java#L256)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Shouldn't we re-throw it instead of creating a new one?
@abosii-harness 
</blockquote></details>

just throw ex, not needed to be wrapped in runtime exception.  As well, throwing in runtime only cause will remove TaskNGDataException from trace and will loose command unit progress
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateBPStep.java` line 176 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBD8VFz1skGAB2aE3ab:-NBD8VFz1skGAB2aE3ac:b9lj2au) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateBPStep.java#L176)):*
> ```Java
>           String.format("Error while processing Azure Create ARM Resource Task response %s", ex.getMessage());
>       log.error(errorMsg, ex);
>       throw new RuntimeException(ex.getCause());
> ```

same here, throw ex

**Review by abosii-harness — APPROVED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 1 change requests. 3. unresolved discussions.

**Review by nassergonzalez — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 1 change requests. 4. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java` line 256 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBCZPmq2_9hW50lgBs4:-NBGvDxn6sfNSyRdSFYY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateARMResourceStep.java#L256)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

just throw ex, not needed to be wrapped in runtime exception.  As well, throwing in runtime only cause will remove TaskNGDataException from trace and will loose command unit progress
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateBPStep.java` line 176 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBD8VFz1skGAB2aE3ab:-NBGvVnd85aPcnr98W7N:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/main/java/io/harness/cdng/provision/azure/AzureCreateBPStep.java#L176)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

same here, throw ex
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/test/resources/recaster/alias/alias.txt` line 466 at r2](https://reviewable.io/reviews/harness/harness-core/36980#-NBCZoF94jvY4a153P9C:-NBH2JFc7l0AJlOYEQ5C:bhuxhay) ([raw file](https://github.com/harness/harness-core/blob/f9ea83e4155ce237b93627a7459d271aabc6ec3b/125-cd-nextgen/src/test/resources/recaster/alias/alias.txt#L466)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Do we need that?
</blockquote></details>

if _that_ was the end carriage i removed it

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 1. unresolved discussions.

**Review by hextechpal — APPROVED** [review]
> Reviewed 5 of 9 files at r1, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 1. unresolved discussions.

**Review by abosii-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: 1. unresolved discussions.

**Review by vlica-harness — APPROVED** [review]
> Reviewed 5 of 9 files at r1, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36980)* status: :shipit: complete! ✓

**Pull request merged**